### ISSUE TITLE
Logging levels are broken

### ISSUE THREAD

**Issue by zachriggle** [Gallopsled/pwntools#380]
This is not the expected behavior.  `"Goodbye"` should be printed after expressly setting the log level on the `Logger` object.

```
>>> from pwn import *
>>> import logging
>>> log.info('Hello')
[*] Hello
>>> log.debug('Goodbye')
>>> log.setLevel(logging.DEBUG)
>>> log.debug('Goodbye')
>>> context.log_level = 'debug'
>>> log.debug('Goodbye')
[DEBUG] Goodbye
```

**Comment by zachriggle** [comment]
Note that it works in the `2.2` release before all of the log reworking.

```
~/pwntools ❯❯❯ git branch -v | grep \*
* (detached from 2.2)       6225901 Merge pull request #343 from Gallopsled/iss-342
```

**Comment by br0ns** [comment]
I don't agree that "Goodbye" should be printed;  the global log level set in `context` is still set to `"INFO"`.  Note that a log record is being generated internally it's just not getting printed.  I think this is the correct behavior.

**Comment by br0ns** [comment]
To elaborate, "Goodbye" is not printed because `context.log_level` doesn't control the level of the loggers but rather the level of the handler.  The benefit of this is that a custom handler can be installed (and generate output) even when a script is silenced (by `context.log_level = 'ERROR'`).

So @IdolfHatler, @kristoff3r and I have just been discussing this issue.  We all agree that the way things currently work are correct.

**Comment by zachriggle** [comment]
This is a departure from the behavior of the logging module.

**Comment by zachriggle** [comment]
Please elaborate as to how one would turn up the verbosity of a single Logger object, e.g. Exactly one tube.

**Comment by zachriggle** [comment]
Part of the context for this is that we used to be able to specify log_level on an individual tube. I have a branch that implements this (tl;dr is that tube subclasses Logger, then replace log.info with self.info).

That functionality is broken with the current mechanism of logging, but would work in 2.2

**Comment by IdolfHatler** [comment]
Every log message should be propagated up the logging tree, no matter the value of `context.log_level`. I can only accept deviations to this if it is for performance reasons. (I even think we are doing this wrong currently as we do some filtering not for performance reasons.)

What you are trying to do is have less propagation and do filtering this way. That IMO is not the right way to do it -- filtering should happen in the handlers.

There is a really good reason for this. For instance, lets say I wanted to download a file:

```py
#!/usr/bin/env python

from pwn import *

s = ssh('mathias', 'ask.diku.dk')
s.download_file('/tmp/fobar')
```

This fails (the file does not exist), but I cannot figure out why or even that it is inside the ssh module (I'm stupid today). If we ever start doing filtering on the individual modules, then I would not be able to log *everything* to a file, (which is nice). To find out what is wrong with this example, I simply do `python doit.py DEBUG LOG_FILE=doit.log`. I then look at it and see what made me confused.

If we had followed your model, then this would not have been possible, because inside ssh.py, there is a `with context.local(log_level = 'ERROR')` around the relevant piece of code.

Bottom line: NO, we should not do filtering of messages, everything should propagate. However I can also see your use-case as very relevant, how about an alternative idea instead. Would you be okay something like this instead?

```py
#!/usr/bin/env python

from pwn import *

sshlog = getLogger('pwnlib.tubes.ssh')
sshlog.addHandler(pwnLogHandler('WARN'))
with context.local(log_level = 'ERROR'):
    do_stuff()
```

**Comment by zachriggle** [comment]
So the conclusion is that the implementation and use pattern of the standard 'logging' library is incorrect?

Zach Riggle

**Comment by IdolfHatler** [comment]
Yes, we agree that the logging module is weird and/or broken.

Would you be okay with my solution?

**Comment by zachriggle** [comment]
One last time, to verify that we're on the same page, I'm discussing this behavior:

```py
import logging

logging.basicConfig()

a = logging.getLogger('pwnlib.a')
aa = logging.getLogger('pwnlib.a.a')
ab = logging.getLogger('pwnlib.a.b')
abx = logging.getLogger('pwnlib.a.b.x')
aby = logging.getLogger('pwnlib.a.b.y')

a.setLevel(logging.WARNING)
ab.setLevel(logging.INFO)
abx.setLevel(logging.DEBUG)

for logger in [a, aa, ab, abx, aby]:
    logger.warning("Warning")
    logger.info("INFO")
    logger.debug("DEBUG")
```

The correct output is:

```
WARNING:pwnlib.a:Warning
WARNING:pwnlib.a.a:Warning
WARNING:pwnlib.a.b:Warning
INFO:pwnlib.a.b:INFO
WARNING:pwnlib.a.b.x:Warning
INFO:pwnlib.a.b.x:INFO
DEBUG:pwnlib.a.b.x:DEBUG
WARNING:pwnlib.a.b.y:Warning
INFO:pwnlib.a.b.y:INFO
```

This works in pwnlib 2.2, by adding `from pwn import *` at the top of the file, and gives the following output:

```
[!] Warning
WARNING:pwnlib.a:Warning
[!] Warning
WARNING:pwnlib.a.a:Warning
[!] Warning
WARNING:pwnlib.a.b:Warning
[*] INFO
INFO:pwnlib.a.b:INFO
[!] Warning
WARNING:pwnlib.a.b.x:Warning
[*] INFO
INFO:pwnlib.a.b.x:INFO
[DEBUG] DEBUG
DEBUG:pwnlib.a.b.x:DEBUG
[!] Warning
WARNING:pwnlib.a.b.y:Warning
[*] INFO
INFO:pwnlib.a.b.y:INFO
```

This does not work in the current master.  Specifically, messages emitted by `a.b.x` at log level `debug` should be shown.  We are incorrectly using a `Handler` to filter these, by making `context.log_level` be a lower bound on the log level.  Note that the default handler (added to the root by `basicConfig()`) correctly prints out the message, but the `pwnlib` handler does not.

```
Warning
WARNING:pwnlib.a:Warning
Warning
WARNING:pwnlib.a.a:Warning
Warning
WARNING:pwnlib.a.b:Warning
INFO
INFO:pwnlib.a.b:INFO
Warning
WARNING:pwnlib.a.b.x:Warning
INFO
INFO:pwnlib.a.b.x:INFO
DEBUG:pwnlib.a.b.x:DEBUG
Warning
WARNING:pwnlib.a.b.y:Warning
INFO
INFO:pwnlib.a.b.y:INFO
```

**Comment by IdolfHatler** [comment]
I can see your use-case. I can see that it is not working with the current master and why. I will here try to explain why what we do is desirable and the reasons why your use-case is not working.

In the end I will explain how we fix the logging module so that your way of using it works.

The reason for all the crap is, we use logging in a slightly weird way inside pwnlib. For instance we do things like this:

```py
with context.local(log_level = 'ERROR'):
    c = self.run(process, tty, wd = wd, timeout = Timeout.default)
    data = c.recvall()
    retcode = c.wait()
    c.close()
    return data, retcode
```

This call to `context.local` is obviously meant to do something. More specifically it signals that whatever goes on in there, we do not want on our screens -- however it might still be valuable debugging, if something really goes wrong.

Now, the question is, what should this call do? There are a few alternatives this means:

1. It affect the filtering inside the logger itself.
2. It can affect the filtering inside the handler.
3. It is a bad example that should be rewritten.

Currently we do 2. (We also do 1, but only for performance reasons. ). We also default to propagate every log-message up the tree. This means that it is possible to add a handler to log these messages to e.g. a file.

You seem to be arguing that we should either do 1 or a combination of 1 and 2. If we did this, then the messages inside that with-block would be forever gone, no matter what we did.

Instead we do the filtering in the handler (which is not "incorrect" -- it is fully supported by the logging module). This means that the *only* semantic of `context.log_level` is what to output to the screen. It does not affect what to put into the logging module, so if you ever want to dump something that does not go to the screen, then you can.

Now, I suggest two fixes for the logging module as it looks now.

1) We should stop monkey patching. The default level of the root pwnlib logger should be 'DEBUG', so that everything will propagate. If you really want to change it, you can -- however there is never any reason to.

The only reason we have that monkey patching is for performance reasons, primarily because of the hexdump inside tubes. We should special case those very few places. They should instead look at context.log_level before calling hexdump.

2) We should make it easier to increase the amount of logging for a specific module (e.g. pwnlib.tubes.ssh). We should do this by adding another handler equivalent to the top-level pwnlib handler. For instance you could do e.g.:

```py
from pwn import *
import logging

logging.getLogger('pwnlib.tubes.ssh').addHandler(pwnTermHandler('WARN'))
```

This will make sure that message of type 'WARN or above, which is in the `pwnlib.tubes.ssh`-tree will be written to the screen, not matter what. br0ns says that he could easily create such a handler and add it to toplevel. We could even do a wrapper function around this behavior if you like.


Conclusion:
You seem to have two things against the way we do things. One is practical: your use-case broken. Second is aesthetical: You think we are using the logging module in the wrong way.

To the first one: Can you accept our solution to your use-case? After that has been implemented, it *is* possible to get extra debugging output from a specific module.

To the second one: No, we are not. Filtering on handlers is supported. We are monkey-patching the log-level for the loggers, but only for performance reason -- and we will stop doing that as soon as we agree on an implementation.

**Comment by zachriggle** [comment]
An easy way to get all of the above is to only use context.log_level when a logger's logLevel is not set.

**Comment by zachriggle** [comment]
Ahh whoops didn't mean to post that or close

**Comment by br0ns** [comment]
So it seems we have come to an agreement on this.  Namely the semantics implemented in #391.

**Comment by zachriggle** [comment]
Nope, current `master` still doesn't work correctly.  Messages 4 and 5 are missing.

```py
from pwn import *
from pwnlib.log import getLogger
import logging

a = getLogger('pwnlib.a')
aa = getLogger('pwnlib.a.a')
ab = getLogger('pwnlib.a.b')
abx = getLogger('pwnlib.a.b.x')
aby = getLogger('pwnlib.a.b.y')

a.setLevel(logging.WARNING)
ab.setLevel(logging.INFO)
abx.setLevel(logging.DEBUG)

for logger in [a, aa, ab, abx, aby]:
    for meth in [logger.warning, logger.info, logger.debug]:
        meth('%s %s' % (logger._logger.name, meth.__name__))

log.debug('NOT OK')
log.info('1')
with context.local(log_level='WARN'):
    assert context.log_level == logging.WARNING
    log.info('NOT OK')
    log.warn('2')
    a.warn('3')
    ab.info('4')
    abx.debug('5')
    aby.debug('NOT OK')

log.info('6')
```

**Comment by zachriggle** [comment]
See #413 which should resolve this, and the problem (https://github.com/Gallopsled/pwntools/pull/386#issuecomment-69434510) that @IdolfHatler found with my previous implementation in #386.
